synapse/storage/_base.py

```diff
@@ -456,6 +456,33 @@ def _new_transaction(
                                 )
                             continue
                     raise
+                finally:
+                    # we're either about to retry with a new cursor, or we're about to
+                    # release the connection. Once we release the connection, it could
+                    # get used for another query, which might do a conn.rollback().
+                    #
+                    # In the latter case, even though that probably wouldn't affect the
+                    # results of this transaction, python's sqlite will reset all
+                    # statements on the connection [1], which will make our cursor
+                    # invalid [2].
+                    #
+                    # While the above probably doesn't apply to postgres, we still need
```

Comment:
We don't actually use cursors in postgres, since we don't name them, therefore we're just fetching all the data on a query. psycopg2's cursor when not named is just an abstraction.

fwiw postgres says that any cursor can see any changes from any other cursor on the same connection, named or not

Reply:
also calling `.close()` on a postgres cursor isn't required as it's implicitly closed at the end of a transaction anyway, as after that it loses semantic meaning. for "client-side cursors" that we use, there's nothing to actually 'close', since no portal was created -- in this case, psycopg2 just sets it as closed and yells if you try and reuse the same cursor.